presto-main/src/main/java/io/prestosql/operator/TaskOutputOperator.java

```diff
@@ -142,14 +142,22 @@ public void addInput(Page page)
 
         page = pagePreprocessor.apply(page);
 
-        List<SerializedPage> serializedPages = splitPage(page, DEFAULT_MAX_PAGE_SIZE_IN_BYTES).stream()
-                .map(serde::serialize)
-                .collect(toImmutableList());
-
-        outputBuffer.enqueue(serializedPages);
+        outputBuffer.enqueue(splitAndSerializePage(page));
         operatorContext.recordOutput(page.getSizeInBytes(), page.getPositionCount());
     }
 
+    private List<SerializedPage> splitAndSerializePage(Page page)
+    {
+        List<Page> split = splitPage(page, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+        ImmutableList.Builder<SerializedPage> builder = ImmutableList.builderWithExpectedSize(split.size());
+        try (PagesSerde.PagesSerdeContext context = serde.newContext()) {
```

Comment:
we are creating temp buffers for a single page, we most likely want them to be long-lived.